{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import pyranges as pr\n",
    "import itertools\n",
    "import sys\n",
    "sys.path.insert(0, '/Genomics/kocherlab/bjarnold/STARRseq/code/notebooks')\n",
    "\n",
    "import functions as fn\n",
    "from collections import defaultdict\n",
    "import os\n",
    "from pyfaidx import Fasta\n",
    "from Bio.Seq import Seq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"Nmel\"\n",
    "bioreps = ['F1', 'F2', 'F3']\n",
    "\n",
    "chr1_test = False\n",
    "qval_threshold = 1 # 1.3 corresponds to pvalue of 0.05, 1 corresponds to pvalue of 0.1\n",
    "macs_folddiff_threshold = 2\n",
    "interval_size = 40\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Genomics/kocherlab/bjarnold/STARRseq/data/peak_calling_snakemake_output/Nmel/MACS2/Nmel-F1_peaks.narrowPeak', '/Genomics/kocherlab/bjarnold/STARRseq/data/peak_calling_snakemake_output/Nmel/MACS2/Nmel-F2_peaks.narrowPeak', '/Genomics/kocherlab/bjarnold/STARRseq/data/peak_calling_snakemake_output/Nmel/MACS2/Nmel-F3_peaks.narrowPeak']\n",
      "['/Genomics/kocherlab/bjarnold/STARRseq/data/peak_calling_snakemake_output/Nmel/MACS2_dedup/Nmel-F1_peaks.narrowPeak', '/Genomics/kocherlab/bjarnold/STARRseq/data/peak_calling_snakemake_output/Nmel/MACS2_dedup/Nmel-F2_peaks.narrowPeak', '/Genomics/kocherlab/bjarnold/STARRseq/data/peak_calling_snakemake_output/Nmel/MACS2_dedup/Nmel-F3_peaks.narrowPeak']\n"
     ]
    }
   ],
   "source": [
    "base_dir = \"/Genomics/kocherlab/bjarnold/STARRseq/data\"\n",
    "wd = f\"{base_dir}/peak_calling_snakemake_output/{species}\"\n",
    "\n",
    "# load MACS peak files\n",
    "macs_raw_peaks_files = fn.get_files(f'{wd}/MACS2', \"*_peaks.narrowPeak\")\n",
    "macs_dedup_peaks_files = fn.get_files(f'{wd}/MACS2_dedup', \"*_peaks.narrowPeak\")\n",
    "\n",
    "print(macs_raw_peaks_files)\n",
    "print(macs_dedup_peaks_files)\n",
    "\n",
    "#load ref genome\n",
    "ref_genome = f'{base_dir}/alternate_references/renamed/{species}.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MACS\n",
      "before filtering: 103313\n",
      "after filtering: 34129\n",
      "before filtering: 101930\n",
      "after filtering: 31722\n",
      "before filtering: 104845\n",
      "after filtering: 36205\n",
      "before filtering: 9645\n",
      "after filtering: 4306\n",
      "before filtering: 13153\n",
      "after filtering: 5499\n",
      "before filtering: 12217\n",
      "after filtering: 4899\n",
      "[34129, 31722, 36205]\n",
      "[4306, 5499, 4899]\n"
     ]
    }
   ],
   "source": [
    "macs_raw_dfs = fn.load_peak_caller_results(macs_raw_peaks_files, chr1_test) # list of macs dataframes\n",
    "macs_dedup_dfs = fn.load_peak_caller_results(macs_dedup_peaks_files, chr1_test) # list of macs dataframes\n",
    "\n",
    "# FILTER PEAK FILES BY SIGNIFICANCE, EFFECT SIZE\n",
    "print('MACS')\n",
    "macs_raw_dfs = fn.filter_by_sig_effect_size(macs_raw_dfs, qval_threshold, macs_folddiff_threshold)\n",
    "macs_dedup_dfs = fn.filter_by_sig_effect_size(macs_dedup_dfs, qval_threshold, macs_folddiff_threshold)\n",
    "\n",
    "print([len(df) for df in macs_raw_dfs])\n",
    "print([len(df) for df in macs_dedup_dfs])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# can't read with pandas because variable number of columns; some rows marked with an asterisk if the repeat overlaps with another higher-scoring repeat\n",
    "# load by reading in all lines instead, then importing into table\n",
    "# then use pyranges to merge overlapping intervals\n",
    "# repeat_pr = fn.load_repeat_modeler_intervals(repeat_file)\n",
    "\n",
    "# open reference genome ref_genome, count the length of each chromosome, and store in a dictionary\n",
    "fasta_sequences = Fasta(ref_genome)\n",
    "\n",
    "scaff_lengths = defaultdict(int)\n",
    "# Iterate over the sequences and print their names and lengths\n",
    "for sequence_name in fasta_sequences.keys():\n",
    "    scaff_lengths[sequence_name] = len(fasta_sequences[sequence_name])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2257796/2266070714.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n",
      "/tmp/ipykernel_2257796/2266070714.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n",
      "/tmp/ipykernel_2257796/2266070714.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n",
      "/tmp/ipykernel_2257796/2266070714.py:29: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n",
      "/tmp/ipykernel_2257796/2266070714.py:29: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n",
      "/tmp/ipykernel_2257796/2266070714.py:29: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n"
     ]
    }
   ],
   "source": [
    "def get_summit_intervals(macs_dfs):\n",
    "    \n",
    "    for m in macs_dfs:\n",
    "        # convert peak/summit coordinate into an interval to overlap with pyranges\n",
    "        m['peak_coord_end'] = m['peak_coord'] + int(interval_size/2)\n",
    "        m['peak_coord_start'] = m['peak_coord'] - int(interval_size/2)\n",
    "\n",
    "    # gather relevant columns, put in new list of dataframes\n",
    "    macs_summit_dfs = [m[['Chromosome', 'peak_coord_start', 'peak_coord_end', 'signalValue']] for m in macs_dfs]\n",
    "\n",
    "    # rename columns to match pyranges format\n",
    "    for m in macs_summit_dfs:\n",
    "        m.columns = ['Chromosome', 'Start', 'End', 'signalValue']\n",
    "        m.reset_index(inplace=True, drop=True) # THIS IS REALLY IMPORTANT, OTHERWISE FETCHING SEQUENCES RETURNS NANS\n",
    "\n",
    "    # convert df to pyranges object\n",
    "    # macs_summit_prs = [pr.PyRanges(m) for m in macs_summit_dfs]\n",
    "\n",
    "    return macs_summit_dfs\n",
    "\n",
    "# extend intervals around summit\n",
    "macs_raw_summit_intervals_df = get_summit_intervals(macs_raw_dfs)\n",
    "macs_dedup_summit_intervals_df = get_summit_intervals(macs_dedup_dfs)\n",
    "\n",
    "# grab sequences from reference genome within intervals\n",
    "for m in macs_raw_summit_intervals_df:\n",
    "    m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n",
    "for m in macs_dedup_summit_intervals_df:\n",
    "    m.loc[:,'Sequence'] = pr.get_sequence(pr.PyRanges(m), ref_genome)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>signalValue</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>NMEL_chr_1</td>\n",
       "      <td>14192628</td>\n",
       "      <td>14192668</td>\n",
       "      <td>3.46729</td>\n",
       "      <td>TCGTCAGATACATAAAATCTTAGCAGTGAGGTTGACGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606</th>\n",
       "      <td>NMEL_chr_1</td>\n",
       "      <td>14197608</td>\n",
       "      <td>14197648</td>\n",
       "      <td>2.38467</td>\n",
       "      <td>AATCAATAACTCAAAAACTACTCAATCAATCTTGCTGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>NMEL_chr_1</td>\n",
       "      <td>14202616</td>\n",
       "      <td>14202656</td>\n",
       "      <td>2.03423</td>\n",
       "      <td>TCTACTTTCAGCGATACCAATGAAACCATATTGTTCGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1608</th>\n",
       "      <td>NMEL_chr_1</td>\n",
       "      <td>14205308</td>\n",
       "      <td>14205348</td>\n",
       "      <td>3.86016</td>\n",
       "      <td>GCGAGCAATAATTTATTTGATTATTGAAAGAAACTTTCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1609</th>\n",
       "      <td>NMEL_chr_1</td>\n",
       "      <td>14212178</td>\n",
       "      <td>14212218</td>\n",
       "      <td>5.24032</td>\n",
       "      <td>GAGCCCTGAATATTTTTTACCATAGCAGAAATAAAAGCCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End  signalValue  \\\n",
       "1605  NMEL_chr_1  14192628  14192668      3.46729   \n",
       "1606  NMEL_chr_1  14197608  14197648      2.38467   \n",
       "1607  NMEL_chr_1  14202616  14202656      2.03423   \n",
       "1608  NMEL_chr_1  14205308  14205348      3.86016   \n",
       "1609  NMEL_chr_1  14212178  14212218      5.24032   \n",
       "\n",
       "                                      Sequence  \n",
       "1605  TCGTCAGATACATAAAATCTTAGCAGTGAGGTTGACGAAA  \n",
       "1606  AATCAATAACTCAAAAACTACTCAATCAATCTTGCTGAAA  \n",
       "1607  TCTACTTTCAGCGATACCAATGAAACCATATTGTTCGAAG  \n",
       "1608  GCGAGCAATAATTTATTTGATTATTGAAAGAAACTTTCTG  \n",
       "1609  GAGCCCTGAATATTTTTTACCATAGCAGAAATAAAAGCCC  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = macs_raw_summit_intervals_df[0]\n",
    "# print(tmp.Chromosome.unique())\n",
    "tmp[tmp['Chromosome'] == 'NMEL_chr_2'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = tmp[['Chromosome', 'Start', 'End']]\n",
    "# write tmp to a tab separated file\n",
    "tmp.to_csv(f'./tmp.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome     0\n",
      "Start          0\n",
      "End            0\n",
      "signalValue    0\n",
      "Sequence       0\n",
      "dtype: int64\n",
      "Chromosome     0\n",
      "Start          0\n",
      "End            0\n",
      "signalValue    0\n",
      "Sequence       0\n",
      "dtype: int64\n",
      "Chromosome     0\n",
      "Start          0\n",
      "End            0\n",
      "signalValue    0\n",
      "Sequence       0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for m in macs_dedup_summit_intervals_df:\n",
    "    print(m.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome     0\n",
      "Start          0\n",
      "End            0\n",
      "signalValue    0\n",
      "Sequence       0\n",
      "dtype: int64\n",
      "Chromosome     0\n",
      "Start          0\n",
      "End            0\n",
      "signalValue    0\n",
      "Sequence       0\n",
      "dtype: int64\n",
      "Chromosome     0\n",
      "Start          0\n",
      "End            0\n",
      "signalValue    0\n",
      "Sequence       0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for m in macs_raw_summit_intervals_df:\n",
    "    print(m.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>signalValue</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4849</th>\n",
       "      <td>NMEL_unplaced_33</td>\n",
       "      <td>9630</td>\n",
       "      <td>9670</td>\n",
       "      <td>2.32957</td>\n",
       "      <td>TTTCATGCTCGCAACACTCGCAAACGATCTTCCATTCATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4850</th>\n",
       "      <td>NMEL_unplaced_35</td>\n",
       "      <td>8963</td>\n",
       "      <td>9003</td>\n",
       "      <td>2.69679</td>\n",
       "      <td>TTCTTACATTCATCCAATCGTCCCTTAAGTCTCTCATTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4851</th>\n",
       "      <td>NMEL_unplaced_38</td>\n",
       "      <td>5687</td>\n",
       "      <td>5727</td>\n",
       "      <td>2.69679</td>\n",
       "      <td>ACGCATCGACCGAATGATTCATTTCGATGATTACTGACCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4852</th>\n",
       "      <td>NMEL_unplaced_393</td>\n",
       "      <td>673</td>\n",
       "      <td>713</td>\n",
       "      <td>3.23615</td>\n",
       "      <td>ATTATGATCCTCTCTCGGCAGTGATGTCGCACTGTCCAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4853</th>\n",
       "      <td>NMEL_unplaced_400</td>\n",
       "      <td>490</td>\n",
       "      <td>530</td>\n",
       "      <td>2.69679</td>\n",
       "      <td>ATGGGTGAATATGAGCAGTGTGTTGTGAAATTGATGTGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4854</th>\n",
       "      <td>NMEL_unplaced_42</td>\n",
       "      <td>11257</td>\n",
       "      <td>11297</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>TGAATCTGTGCAGCGGGAACGACCTCGGACTACCAGGGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4855</th>\n",
       "      <td>NMEL_unplaced_439</td>\n",
       "      <td>1720</td>\n",
       "      <td>1760</td>\n",
       "      <td>2.87658</td>\n",
       "      <td>ACCTTCTCTCCTTGGGCAAATGAGATCGGCCGTGCACGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4856</th>\n",
       "      <td>NMEL_unplaced_443</td>\n",
       "      <td>838</td>\n",
       "      <td>878</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>CCGAGAAGCGACCTGGATGTGTGTTATGCGAGTTCACTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4857</th>\n",
       "      <td>NMEL_unplaced_443</td>\n",
       "      <td>1977</td>\n",
       "      <td>2017</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>GCGATCTGGATGTGTGTTATGCGAGTACACATTGATCTTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4858</th>\n",
       "      <td>NMEL_unplaced_447</td>\n",
       "      <td>2452</td>\n",
       "      <td>2492</td>\n",
       "      <td>3.77551</td>\n",
       "      <td>TCGTAGCGGTCTTTGAACGCGAGGCCCCGCCCCGGATAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4859</th>\n",
       "      <td>NMEL_unplaced_448</td>\n",
       "      <td>1532</td>\n",
       "      <td>1572</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>CCACTAAGTTCTCAGTCGTATTATCTTAATATTCAATGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4860</th>\n",
       "      <td>NMEL_unplaced_470</td>\n",
       "      <td>1921</td>\n",
       "      <td>1961</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>AAGTGTACTCGTATAACCCAAATCCAGATCGCTTCGCGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4861</th>\n",
       "      <td>NMEL_unplaced_48</td>\n",
       "      <td>3043</td>\n",
       "      <td>3083</td>\n",
       "      <td>2.22375</td>\n",
       "      <td>AGCCATAAGGCCGAGAAGCGATCTGGATGAGTGTTATGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4862</th>\n",
       "      <td>NMEL_unplaced_480</td>\n",
       "      <td>572</td>\n",
       "      <td>612</td>\n",
       "      <td>4.13508</td>\n",
       "      <td>TGGTGATATCTCGGGTGCTATTAGAGATAGCGCAATAAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4863</th>\n",
       "      <td>NMEL_unplaced_484</td>\n",
       "      <td>627</td>\n",
       "      <td>667</td>\n",
       "      <td>2.63649</td>\n",
       "      <td>TCACACGCAAATCTCACACTCCTGATAAACTCATATGTAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4864</th>\n",
       "      <td>NMEL_unplaced_505</td>\n",
       "      <td>1370</td>\n",
       "      <td>1410</td>\n",
       "      <td>2.45271</td>\n",
       "      <td>CGTCATTCTACCATTCACATTCATTGTACTCAGTCGCTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4865</th>\n",
       "      <td>NMEL_unplaced_52</td>\n",
       "      <td>6541</td>\n",
       "      <td>6581</td>\n",
       "      <td>3.67907</td>\n",
       "      <td>CTATCTCTAATAGCACCCGAGATATCACCAGATATATATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4866</th>\n",
       "      <td>NMEL_unplaced_526</td>\n",
       "      <td>1749</td>\n",
       "      <td>1789</td>\n",
       "      <td>2.71357</td>\n",
       "      <td>GGGTCGGACACTCAGGGTCAGGGAAAAATCCAATATAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4867</th>\n",
       "      <td>NMEL_unplaced_55</td>\n",
       "      <td>6528</td>\n",
       "      <td>6568</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>GTCTCTTTCACTCTACATATCTTTTTCTTATACTCACTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4868</th>\n",
       "      <td>NMEL_unplaced_555</td>\n",
       "      <td>817</td>\n",
       "      <td>857</td>\n",
       "      <td>4.49465</td>\n",
       "      <td>AATGTACGCGAATAACCCAAATCCAGATCGCTCCTCGGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4869</th>\n",
       "      <td>NMEL_unplaced_580</td>\n",
       "      <td>1991</td>\n",
       "      <td>2031</td>\n",
       "      <td>3.05636</td>\n",
       "      <td>ATTGTGCTAAACATTCGTAGTATGTAAAAGTATTAGCTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4870</th>\n",
       "      <td>NMEL_unplaced_585</td>\n",
       "      <td>1169</td>\n",
       "      <td>1209</td>\n",
       "      <td>4.67444</td>\n",
       "      <td>ATACTTTCGCTCATTCACACGCAACGTCGGATCTCTCACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4871</th>\n",
       "      <td>NMEL_unplaced_59</td>\n",
       "      <td>6044</td>\n",
       "      <td>6084</td>\n",
       "      <td>2.87658</td>\n",
       "      <td>CAGCGGGAACGACCTCGGACCACCAGGGGCCGAGAAAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4872</th>\n",
       "      <td>NMEL_unplaced_599</td>\n",
       "      <td>2077</td>\n",
       "      <td>2117</td>\n",
       "      <td>3.23615</td>\n",
       "      <td>GACCACCAGGGGCCGAGAAAACCACAGTCTCATGGGGCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4873</th>\n",
       "      <td>NMEL_unplaced_618</td>\n",
       "      <td>1251</td>\n",
       "      <td>1291</td>\n",
       "      <td>3.23615</td>\n",
       "      <td>TCGTCCCTATGAGACATCGTCCTGGAGTCCCATCACTGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4874</th>\n",
       "      <td>NMEL_unplaced_633</td>\n",
       "      <td>723</td>\n",
       "      <td>763</td>\n",
       "      <td>2.87658</td>\n",
       "      <td>ACAATCAGGGTCAGGGAAAAATCCAATATATATCTGTTGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4875</th>\n",
       "      <td>NMEL_unplaced_640</td>\n",
       "      <td>1606</td>\n",
       "      <td>1646</td>\n",
       "      <td>2.71006</td>\n",
       "      <td>CATTCAACCTCTCGATCACTCTCTTCCATTCTAAATTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4876</th>\n",
       "      <td>NMEL_unplaced_651</td>\n",
       "      <td>1833</td>\n",
       "      <td>1873</td>\n",
       "      <td>2.45271</td>\n",
       "      <td>AGTTTTATATAGTTTTATATAGTTTTATACAGTTTTATAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4877</th>\n",
       "      <td>NMEL_unplaced_68</td>\n",
       "      <td>1649</td>\n",
       "      <td>1689</td>\n",
       "      <td>2.87658</td>\n",
       "      <td>CGGACCACCAGGGGCCGAGAAAACCACCGACACATGGGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4878</th>\n",
       "      <td>NMEL_unplaced_686</td>\n",
       "      <td>1462</td>\n",
       "      <td>1502</td>\n",
       "      <td>2.62791</td>\n",
       "      <td>ATCTGTGCAGCGGGAACGACCTCGGACCACCAGGGGCCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4879</th>\n",
       "      <td>NMEL_unplaced_718</td>\n",
       "      <td>1731</td>\n",
       "      <td>1771</td>\n",
       "      <td>2.87658</td>\n",
       "      <td>ATCAAAGTGCATTGTCTGTAGGAGGCCCTGGCGCATGAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4880</th>\n",
       "      <td>NMEL_unplaced_74</td>\n",
       "      <td>6071</td>\n",
       "      <td>6111</td>\n",
       "      <td>2.69679</td>\n",
       "      <td>CTCGCGCGCGATACGCGTCATCCTCTTTTTTCCCCCATTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4881</th>\n",
       "      <td>NMEL_unplaced_762</td>\n",
       "      <td>1809</td>\n",
       "      <td>1849</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>TATAAGCGTGGTTTGAGTGAGACTGAAAGATGTATGTGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4882</th>\n",
       "      <td>NMEL_unplaced_769</td>\n",
       "      <td>498</td>\n",
       "      <td>538</td>\n",
       "      <td>2.32816</td>\n",
       "      <td>AGCCTCACCACACATACACAACGCACTTTCACTCAATCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4883</th>\n",
       "      <td>NMEL_unplaced_780</td>\n",
       "      <td>998</td>\n",
       "      <td>1038</td>\n",
       "      <td>2.15835</td>\n",
       "      <td>GGGTCAGGGAAAAATCCAATATATATCTGGTGATATCTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4884</th>\n",
       "      <td>NMEL_unplaced_782</td>\n",
       "      <td>510</td>\n",
       "      <td>550</td>\n",
       "      <td>4.67444</td>\n",
       "      <td>CTCATGGGGCCGGACACTCAGGGTCAGGGAAAAATCCAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4885</th>\n",
       "      <td>NMEL_unplaced_789</td>\n",
       "      <td>981</td>\n",
       "      <td>1021</td>\n",
       "      <td>2.43201</td>\n",
       "      <td>TCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886</th>\n",
       "      <td>NMEL_unplaced_833</td>\n",
       "      <td>1773</td>\n",
       "      <td>1813</td>\n",
       "      <td>3.59572</td>\n",
       "      <td>CACCGATTCGTCATCCATAAATTTAGCGGTGCTTGATAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4887</th>\n",
       "      <td>NMEL_unplaced_840</td>\n",
       "      <td>1221</td>\n",
       "      <td>1261</td>\n",
       "      <td>2.87658</td>\n",
       "      <td>GCATTGTCTGCAGGAGGCCCTGGCGCATGAAACTGTGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4888</th>\n",
       "      <td>NMEL_unplaced_856</td>\n",
       "      <td>1090</td>\n",
       "      <td>1130</td>\n",
       "      <td>3.15229</td>\n",
       "      <td>TGCTATTAGAGATAGCGCAATAAATCAAAGTGCATTGTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4889</th>\n",
       "      <td>NMEL_unplaced_875</td>\n",
       "      <td>405</td>\n",
       "      <td>445</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>GGGCCTCTCTCTCTCTTTCTCCCTCGAACGTAATCAACGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4890</th>\n",
       "      <td>NMEL_unplaced_876</td>\n",
       "      <td>1585</td>\n",
       "      <td>1625</td>\n",
       "      <td>2.68596</td>\n",
       "      <td>TATGCCTCTTGAGGCATATCCGAGGTCACTGCACTCACAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4891</th>\n",
       "      <td>NMEL_unplaced_885</td>\n",
       "      <td>1027</td>\n",
       "      <td>1067</td>\n",
       "      <td>3.95529</td>\n",
       "      <td>ACGATGCCTGGAAACAAATAGAATCCGGCGGCACTTACGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4892</th>\n",
       "      <td>NMEL_unplaced_895</td>\n",
       "      <td>1046</td>\n",
       "      <td>1086</td>\n",
       "      <td>3.52875</td>\n",
       "      <td>CTCATAGGGACCAAGGCTACATCCGGCTGACCGATGGGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4893</th>\n",
       "      <td>NMEL_unplaced_907</td>\n",
       "      <td>1653</td>\n",
       "      <td>1693</td>\n",
       "      <td>2.69679</td>\n",
       "      <td>CACGGAACAAGCAATCTATCTAGTCACGCGTGTTCAACGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4894</th>\n",
       "      <td>NMEL_unplaced_910</td>\n",
       "      <td>418</td>\n",
       "      <td>458</td>\n",
       "      <td>2.05682</td>\n",
       "      <td>TCTGGATGTGTGTTATTCGAGTACACTTTGATCTTAGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4895</th>\n",
       "      <td>NMEL_unplaced_92</td>\n",
       "      <td>2430</td>\n",
       "      <td>2470</td>\n",
       "      <td>2.88317</td>\n",
       "      <td>ACAGATAGTGCAATAAATCAAAGTGCATTGTCTGTAGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4896</th>\n",
       "      <td>NMEL_unplaced_951</td>\n",
       "      <td>611</td>\n",
       "      <td>651</td>\n",
       "      <td>2.30851</td>\n",
       "      <td>TGGCGCCGGACACTCAGGGTCAGGGAAAAATCCAATATAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4897</th>\n",
       "      <td>NMEL_unplaced_969</td>\n",
       "      <td>1400</td>\n",
       "      <td>1440</td>\n",
       "      <td>2.51701</td>\n",
       "      <td>TATCTCGGGTGCTATTAGAGATAGCGCAATAAATCAAAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4898</th>\n",
       "      <td>NMEL_unplaced_995</td>\n",
       "      <td>465</td>\n",
       "      <td>505</td>\n",
       "      <td>3.75767</td>\n",
       "      <td>ACAAACATCCAGAACGCTTCTCGGCCTTATGGCTAAGATC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Chromosome  Start    End  signalValue  \\\n",
       "4849   NMEL_unplaced_33   9630   9670      2.32957   \n",
       "4850   NMEL_unplaced_35   8963   9003      2.69679   \n",
       "4851   NMEL_unplaced_38   5687   5727      2.69679   \n",
       "4852  NMEL_unplaced_393    673    713      3.23615   \n",
       "4853  NMEL_unplaced_400    490    530      2.69679   \n",
       "4854   NMEL_unplaced_42  11257  11297      2.51701   \n",
       "4855  NMEL_unplaced_439   1720   1760      2.87658   \n",
       "4856  NMEL_unplaced_443    838    878      2.51701   \n",
       "4857  NMEL_unplaced_443   1977   2017      2.51701   \n",
       "4858  NMEL_unplaced_447   2452   2492      3.77551   \n",
       "4859  NMEL_unplaced_448   1532   1572      2.51701   \n",
       "4860  NMEL_unplaced_470   1921   1961      2.51701   \n",
       "4861   NMEL_unplaced_48   3043   3083      2.22375   \n",
       "4862  NMEL_unplaced_480    572    612      4.13508   \n",
       "4863  NMEL_unplaced_484    627    667      2.63649   \n",
       "4864  NMEL_unplaced_505   1370   1410      2.45271   \n",
       "4865   NMEL_unplaced_52   6541   6581      3.67907   \n",
       "4866  NMEL_unplaced_526   1749   1789      2.71357   \n",
       "4867   NMEL_unplaced_55   6528   6568      2.51701   \n",
       "4868  NMEL_unplaced_555    817    857      4.49465   \n",
       "4869  NMEL_unplaced_580   1991   2031      3.05636   \n",
       "4870  NMEL_unplaced_585   1169   1209      4.67444   \n",
       "4871   NMEL_unplaced_59   6044   6084      2.87658   \n",
       "4872  NMEL_unplaced_599   2077   2117      3.23615   \n",
       "4873  NMEL_unplaced_618   1251   1291      3.23615   \n",
       "4874  NMEL_unplaced_633    723    763      2.87658   \n",
       "4875  NMEL_unplaced_640   1606   1646      2.71006   \n",
       "4876  NMEL_unplaced_651   1833   1873      2.45271   \n",
       "4877   NMEL_unplaced_68   1649   1689      2.87658   \n",
       "4878  NMEL_unplaced_686   1462   1502      2.62791   \n",
       "4879  NMEL_unplaced_718   1731   1771      2.87658   \n",
       "4880   NMEL_unplaced_74   6071   6111      2.69679   \n",
       "4881  NMEL_unplaced_762   1809   1849      2.51701   \n",
       "4882  NMEL_unplaced_769    498    538      2.32816   \n",
       "4883  NMEL_unplaced_780    998   1038      2.15835   \n",
       "4884  NMEL_unplaced_782    510    550      4.67444   \n",
       "4885  NMEL_unplaced_789    981   1021      2.43201   \n",
       "4886  NMEL_unplaced_833   1773   1813      3.59572   \n",
       "4887  NMEL_unplaced_840   1221   1261      2.87658   \n",
       "4888  NMEL_unplaced_856   1090   1130      3.15229   \n",
       "4889  NMEL_unplaced_875    405    445      2.51701   \n",
       "4890  NMEL_unplaced_876   1585   1625      2.68596   \n",
       "4891  NMEL_unplaced_885   1027   1067      3.95529   \n",
       "4892  NMEL_unplaced_895   1046   1086      3.52875   \n",
       "4893  NMEL_unplaced_907   1653   1693      2.69679   \n",
       "4894  NMEL_unplaced_910    418    458      2.05682   \n",
       "4895   NMEL_unplaced_92   2430   2470      2.88317   \n",
       "4896  NMEL_unplaced_951    611    651      2.30851   \n",
       "4897  NMEL_unplaced_969   1400   1440      2.51701   \n",
       "4898  NMEL_unplaced_995    465    505      3.75767   \n",
       "\n",
       "                                      Sequence  \n",
       "4849  TTTCATGCTCGCAACACTCGCAAACGATCTTCCATTCATA  \n",
       "4850  TTCTTACATTCATCCAATCGTCCCTTAAGTCTCTCATTCT  \n",
       "4851  ACGCATCGACCGAATGATTCATTTCGATGATTACTGACCG  \n",
       "4852  ATTATGATCCTCTCTCGGCAGTGATGTCGCACTGTCCAGA  \n",
       "4853  ATGGGTGAATATGAGCAGTGTGTTGTGAAATTGATGTGTG  \n",
       "4854  TGAATCTGTGCAGCGGGAACGACCTCGGACTACCAGGGGC  \n",
       "4855  ACCTTCTCTCCTTGGGCAAATGAGATCGGCCGTGCACGGG  \n",
       "4856  CCGAGAAGCGACCTGGATGTGTGTTATGCGAGTTCACTTT  \n",
       "4857  GCGATCTGGATGTGTGTTATGCGAGTACACATTGATCTTA  \n",
       "4858  TCGTAGCGGTCTTTGAACGCGAGGCCCCGCCCCGGATAGA  \n",
       "4859  CCACTAAGTTCTCAGTCGTATTATCTTAATATTCAATGTT  \n",
       "4860  AAGTGTACTCGTATAACCCAAATCCAGATCGCTTCGCGGC  \n",
       "4861  AGCCATAAGGCCGAGAAGCGATCTGGATGAGTGTTATGCG  \n",
       "4862  TGGTGATATCTCGGGTGCTATTAGAGATAGCGCAATAAAC  \n",
       "4863  TCACACGCAAATCTCACACTCCTGATAAACTCATATGTAA  \n",
       "4864  CGTCATTCTACCATTCACATTCATTGTACTCAGTCGCTCG  \n",
       "4865  CTATCTCTAATAGCACCCGAGATATCACCAGATATATATT  \n",
       "4866  GGGTCGGACACTCAGGGTCAGGGAAAAATCCAATATAAAT  \n",
       "4867  GTCTCTTTCACTCTACATATCTTTTTCTTATACTCACTCA  \n",
       "4868  AATGTACGCGAATAACCCAAATCCAGATCGCTCCTCGGCC  \n",
       "4869  ATTGTGCTAAACATTCGTAGTATGTAAAAGTATTAGCTAG  \n",
       "4870  ATACTTTCGCTCATTCACACGCAACGTCGGATCTCTCACA  \n",
       "4871  CAGCGGGAACGACCTCGGACCACCAGGGGCCGAGAAAACC  \n",
       "4872  GACCACCAGGGGCCGAGAAAACCACAGTCTCATGGGGCCG  \n",
       "4873  TCGTCCCTATGAGACATCGTCCTGGAGTCCCATCACTGCT  \n",
       "4874  ACAATCAGGGTCAGGGAAAAATCCAATATATATCTGTTGA  \n",
       "4875  CATTCAACCTCTCGATCACTCTCTTCCATTCTAAATTATC  \n",
       "4876  AGTTTTATATAGTTTTATATAGTTTTATACAGTTTTATAT  \n",
       "4877  CGGACCACCAGGGGCCGAGAAAACCACCGACACATGGGGC  \n",
       "4878  ATCTGTGCAGCGGGAACGACCTCGGACCACCAGGGGCCGA  \n",
       "4879  ATCAAAGTGCATTGTCTGTAGGAGGCCCTGGCGCATGAAT  \n",
       "4880  CTCGCGCGCGATACGCGTCATCCTCTTTTTTCCCCCATTG  \n",
       "4881  TATAAGCGTGGTTTGAGTGAGACTGAAAGATGTATGTGTG  \n",
       "4882  AGCCTCACCACACATACACAACGCACTTTCACTCAATCCA  \n",
       "4883  GGGTCAGGGAAAAATCCAATATATATCTGGTGATATCTCG  \n",
       "4884  CTCATGGGGCCGGACACTCAGGGTCAGGGAAAAATCCAAT  \n",
       "4885  TCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAG  \n",
       "4886  CACCGATTCGTCATCCATAAATTTAGCGGTGCTTGATAGC  \n",
       "4887  GCATTGTCTGCAGGAGGCCCTGGCGCATGAAACTGTGCAG  \n",
       "4888  TGCTATTAGAGATAGCGCAATAAATCAAAGTGCATTGTCT  \n",
       "4889  GGGCCTCTCTCTCTCTTTCTCCCTCGAACGTAATCAACGT  \n",
       "4890  TATGCCTCTTGAGGCATATCCGAGGTCACTGCACTCACAA  \n",
       "4891  ACGATGCCTGGAAACAAATAGAATCCGGCGGCACTTACGA  \n",
       "4892  CTCATAGGGACCAAGGCTACATCCGGCTGACCGATGGGAA  \n",
       "4893  CACGGAACAAGCAATCTATCTAGTCACGCGTGTTCAACGC  \n",
       "4894  TCTGGATGTGTGTTATTCGAGTACACTTTGATCTTAGCCA  \n",
       "4895  ACAGATAGTGCAATAAATCAAAGTGCATTGTCTGTAGGAG  \n",
       "4896  TGGCGCCGGACACTCAGGGTCAGGGAAAAATCCAATATAT  \n",
       "4897  TATCTCGGGTGCTATTAGAGATAGCGCAATAAATCAAAGT  \n",
       "4898  ACAAACATCCAGAACGCTTCTCGGCCTTATGGCTAAGATC  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "macs_dedup_summit_intervals_df[2].tail(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8404"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scaff_lengths['NMEL_unplaced_52']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome     0\n",
       "Start          0\n",
       "End            0\n",
       "signalValue    0\n",
       "Sequence       0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "macs_dedup_summit_intervals_df[2].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = macs_dedup_summit_intervals_df[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# kmer_df['sequence'] = pr.get_sequence(pr.PyRanges(kmer_df), ref_genome)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datascience",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
